### DISCUSSION TITLE
chapter_greedy/max_capacity_problem/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#640] [Announcements]
# chapter_greedy/max_capacity_problem/

动画图解、一键运行的数据结构与算法教程

https://www.hello-algo.com/chapter_greedy/max_capacity_problem/

**Comment by Junyu-Tang** [comment]
啊，让我想到了LeetCode上的《接雨水》的问题...感觉有点像。

**Reply by nuist-zjx** [reply]
是这样，双指针

**Reply by applejwjcat** [reply]
接雨水要怎么做

**Reply by syutengu** [reply]
lc#11. Container With Most Water

**Comment by RoyLuoNanjing** [comment]
图画的用心，赞一个

**Comment by Terran0629** [comment]
牛逼啊K神，醍醐灌顶

**Comment by LeoY2020** [comment]
为啥要第`1.`步、第`2.`步？

**Reply by dqqql** [reply]
啥意思，你具体指哪里？

**Comment by xuanxuanxie** [comment]
我的猪猪脑子很难得出需要移动短板这一个策略TAT

**Reply by BigHeadUltra** [reply]
宝贝，你太可爱了TAT

**Reply by fufu39** [reply]
哈哈哈

**Reply by YF-SA** [reply]
终于找到你~~~

**Comment by June3426** [comment]
我对这些精美的图例毫无抵抗力！大爱K神！

**Comment by gonisan** [comment]
这思路确实巧妙

**Comment by SSR-chen** [comment]
一句话醍醐灌顶啊，膜拜

**Comment by AL-persist** [comment]
不是很懂“这是因为在移动长板j后，宽度j-i肯定变小；而高度由短板决定，因此高度只可能不变（i仍为短板）或变小（移动后的j成为短板）”这句话，假如移动长板j之后，j还仍可能时长坂吧，甚至有可能高度比原来的都大

**Reply by poisonyu** [reply]
假如移动长板j之后，如果j还是长板，此时高度不取j对应的板高度，而是取i（短板）对应的板高度，此时高度与移动前相等

**Reply by wdllmh** [reply]
`min(i, j) <= i` 无论长板j怎么变，所以`V = min(i, j) * (j - i) <= i*(j-i)`, 这里的`i*(j-i)`在j的最初时最大，所以跳过的都是小的

**Reply by superfeimao** [reply]
面积取决于短板*宽度

**Reply by njllljhfh** [reply]
i:是短板，j:是长板
移动 j 后有2种情况，1.相对短的板子长度不变，2.相对短的板子长度变短。所以结果一定是容量变小。

**Comment by Jacobinwwey** [comment]
此处不理解的是：既然内移会降低面积，那么移动短板后比较[i,j]内没有比短板更高的板就可以停止移动了吧？为什么要让短板穷举呢？

**Reply by Jacobinwwey** [reply]
是数学逻辑为算法效率妥协的结果吗？比较一遍是否具有更高板子最差复杂度要O(n)，但穷举只需要O(1)？

**Reply by LeiSteins** [reply]
内移不一定会降低面积

**Reply by SenyFish** [reply]
文章最后说了“每轮的贪心选择都会“跳过”一些状态”，按你的思路就不是贪心了，时间复杂度会达到O（n^2）
